### PR TITLE
🔧Use mypy for typechecking

### PR DESCRIPTION
**Pull request by s-weigand**
For the `Decorated property not supported` issue, refer to https://github.com/python/mypy/issues/1362
Locally you can run this check with `pre-commit run -a mypy`, this will currently give you the 233 error the CI shows.

### CODE REVIEW

**Comment by s-weigand** [comment]
Current error count:
`Found 231 errors in 41 files (checked 100 source files)`

Some helpful resources are:
https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
https://github.com/python/typing/issues/553

Also, we might need to create stub files for classes with dynamically added properties.
https://mypy.readthedocs.io/en/stable/stubs.html
E.g.
```
glotaran/builtin/models/kinetic_spectrum/spectral_shape.py:34: error: "SpectralShapeGaussian" has no attribute "amplitude"
glotaran/builtin/models/kinetic_spectrum/spectral_shape.py:35: error: "SpectralShapeGaussian" has no attribute "location"
glotaran/builtin/models/kinetic_spectrum/spectral_shape.py:35: error: "SpectralShapeGaussian" has no attribute "width"
```

https://github.com/glotaran/pyglotaran/blob/81be7b08bc1bfe7a571c4cea21cdf72734a530dd/glotaran/builtin/models/kinetic_spectrum/spectral_shape.py#L10-L18

Or maybe you can find a better way to tell mypy that properties were dynamically added.

Also to avoid `error: Module not callable` some files should be renamed
E.g. `model_attribute.py` can be `attribute.py`, since it is in the package `model` the prefix is redundant.
The whole problem here is that when python starts loading `glotaran/model/__init__.py` at the start `glotaran.model.model_attribute` refers to the module `glotaran/model/model_attribute.py` which is overwritten and after refers to what formerly was `glotaran.model.model_attribute.model_attribute`.

https://github.com/glotaran/pyglotaran/blob/81be7b08bc1bfe7a571c4cea21cdf72734a530dd/glotaran/model/__init__.py#L7-L29

**Comment by s-weigand** [comment]
Bonus tipp you can reproduce [what the pre-commit hook does](https://github.com/pre-commit/mirrors-mypy/blob/master/.pre-commit-hooks.yaml) by running
```console
$ mypy --ignore-missing-imports --scripts-are-modules glotaran/
```

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1458384107)
Found 199 errors in 32 files (checked 100 source files)
After rebasing on #453

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1462988697)
Found 183 errors in 32 files (checked 100 source files)
After rebasing with merged #456

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1483314378)
Found 334 errors in 31 files (checked 85 source files)
After rebasing with merged #442 

Note that mypy is currently configured to only pick up errors when typed parts of the code (which isn't the case everywhere) interact, so adding more typed code might cause more errors, which doesn't necessarily reflect the quality of the code or the typing.

File: setup.cfg [hunk 1]
```diff
@@ -22,3 +22,8 @@ ignore_directives = autoattribute,autoclass,autoexception,autofunction,autometho
 
 [darglint]
 docstring_style = numpy
+
+[mypy]
+ignore_missing_imports = True
+scripts_are_modules = True
+show_error_codes = True
```
Comment (jsnel) [inline setup.cfg:29]: What is the motivation for each of these options?

Comment (s-weigand) [reply setup.cfg:29]: - `ignore_missing_imports`: Prevents errors where dependencies cause errors due to their missing type.
E.g,` Skipping analyzing 'numpy': found module but no type hints or library stubs  [import]`
- `show_error_codes`: Displays the error code at the end of an error, which makes it easier to look it up or if needed ignore an error in a fine-grain manner, e.g. `import` in the above example
- `scripts_are_modules`: Was taken from the [default arguments of the hook](https://github.com/pre-commit/mirrors-mypy/blob/6fd7e2a39e5a8f7c7b14ddcc5290ef69b0c446db/.pre-commit-hooks.yaml#L6) and [does only have an effect on script files without `*.py` ending](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-scripts-are-modules). So I guess we can drop it

Comment (s-weigand) [reply setup.cfg:29]: Also, we might want to add the config in a separate PR so we don't need to pass all the flags to `mypy` when fixing existing errors before we can use the pre-commit hook.

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1489368914)
Found 229 errors in 22 files (checked 85 source files)
After rebasing with merged #467

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1503760934)
Found 230 errors in 23 files (checked 85 source files)
After rebasing with merged #473

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/runs/1749237133)
Found 229 errors in 27 files (checked 86 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1849809062) c72e7b859067dcc10669576a0efc3416baa93f46
Found 284 errors in 25 files (checked 86 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1893165824) 33ff828ab8c65742b01512fe3b24fd10faae1c2a
Found 229 errors in 28 files (checked 93 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1942091785) 40274921c65d8486e6bca0919340d9dd5f4c77db
Found 231 errors in 28 files (checked 93 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=1942456036) 61e2fe94360c33202dba6f482402965284d8c30b
Found 247 errors in 28 files (checked 103 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2051428472) f214d78d6d68d624effad4f2bdaab41b573b866e
Found 247 errors in 28 files (checked 103 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2064873566) c26f97390d3ff088ad350594e3f5ee74638e3d4b
Found 246 errors in 26 files (checked 103 source files)

**Review by jsnel — APPROVED** [review]
👍

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2102859397) 08dae4bf34f9ed9ec643033a4e199d52bb22c540
Found 246 errors in 26 files (checked 104 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2143486751) a7bc0288526956c6f07f87d724231adb6a8317b9
Found 244 errors in 25 files (checked 110 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2249444420) 032cf235eb8431c920da54a6d47e7e58e493abdb
Found 244 errors in 25 files (checked 114 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2264791189) bb3702a9496616b20dff21897acde3870ea58f8a
Found 244 errors in 25 files (checked 114 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2265477798) f4fcbe148dc834a2be24a46f9688f9797a54fa98
Found 246 errors in 26 files (checked 127 source files)

#631 added the following error, which is due to wrong typing of label in ParameterGroup, we should ensure that label is always a string even if generated from a list or loosen the typing.
```
glotaran\deprecation\modules\test\test_glotaran_root.py:114: error: Argument 1 to "get" of "ParameterGroup" has incompatible type "int"; expected "str"  [arg-type]
```

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2272086210) 23741f684795070b8b44bfb5ce3a28994dc44be3
Found 239 errors in 26 files (checked 127 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2631289743) 2a4a8a25abece7514c95044cd61cdb96c989933c
Found 240 errors in 27 files (checked 127 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2663353071) 3a40f6eda26e74b62155343b5ed9481c6801243e
Found 240 errors in 27 files (checked 127 source files)

**Comment by s-weigand** [comment]
Squashed the commits so it is easier to just cherry-pick it over to a branch (e.g. PR branch) and run it there.

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2698618809) a3bf9a248aaaa2a0f54034a51c8374c1abee8d9e
Found 268 errors in 35 files (checked 135 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2700787058) 505248e47dadfce92f548cc5904daecb275b80c9
Found 268 errors in 35 files (checked 135 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2705860781) 968a668fe2f40bd2141c8a7f084b556472bafb4c
Found 235 errors in 33 files (checked 138 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2786968678) cc168aeb026a01583ef2e63b015c56c8abfe9904
Found 252 errors in 35 files (checked 142 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2809555605) b1c3042e21d2ba209fddbd3022d230bddbb4be2f
Found 252 errors in 35 files (checked 142 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2921158832) dc00e6da7b0f598dad9dd24ad490e063ddad9115
Found 308 errors in 36 files (checked 143 source files)

**Comment by s-weigand** [comment]
[Current error count:](https://github.com/glotaran/pyglotaran/pull/444/checks?check_run_id=2921170999) 514a7451887b5d18b7d46e213003a0a073703451
Found 308 errors in 36 files (checked 143 source files)

**Comment by s-weigand** [comment]
This can be closed now

**Pull request closed without merging**